rfcs/45-stageless.md

```diff
@@ -0,0 +1,866 @@
+        .add_system(gravity.label(Physics::Forces))
+        // These systems have a linear chain of ordering dependencies between them
+        // Systems earlier in the chain must run before those later in the chain
+        // Other systems can run in between these systems
+        .add_systems((broad_pass, narrow_pass).chain().label(Physics::CollisionDetection))
+        // Add multiple systems as once to reduce boilerplate!
+        .add_systems((compute_forces, collision_damage).label(Physics::CollisionHandling));
+    }
+}
+```
+
+Labels cannot themselves be labelled: composition is generally clearer than ad-hoc inheritance, and this allows us to compute the configuration for each system more quickly.
+If you need to copy the behavior of a label, use `.configure_label(TargetLabel.configure(SourceLabel.config()))`.
+If you need to ensure that your systems are treated as if they were nested within another label to ensure that other systems have the correct relative ordering, also add that label to your systems.
+
+### System ordering
+
+The most important way we can configure systems is by telling the scheduler *when* they should be run.
+The ordering of systems is always defined relative to other systems: either directly or by checking the systems that belong to a label.
+
+Applying an ordering constraint to or from a label causes a ordering constraint to be created between all individual members of that label.
+If an ordering is defined relative to a non-existent system or an unused label, it will have no effect, emitting a warning.
+This relatively gentle failure mode is important to ensure that plugins can order their systems with relatively strong assumptions that the default system labels exist, but continue to (mostly) work if those systems or labels are not present.
+
+In addition to the `.before` and `.after` methods, you can use **system chains** to create very simple linear dependencies between the successive members of an array of systems.
+(Note to readers: this is not the same as "system chaining" in Bevy 0.6 and earlier: that concept has been renamed to "system piping".)
+
+```rust
+fn main(){
+   App::new()
+   // This systems are connected using a string of ordering constraints
+   .add_systems((compute_attack, 
+                 compute_defense,
+                 check_for_crits,
+                 compute_damage,
+                 deal_damage,
+                 check_for_death)
+                 .chain()
+                 // We can configure and label all systems in the chain at once
+                 .label(GameLabel::Combat)
+   )
+   .run()
+}
+```
+
+#### Ordering with `Commands`
+
+Commands (commonly used to spawn and despawn entities or add and remove components) do not take effect immediately.
+Instead, they are applied whenever the `flush_commands` system runs.
+This exclusive system collects all created commands and applies them to the `World`, mutating it directly.

```

Comment:
I'm not sure how I feel about the `flush_commands` system acting on all pending commands. I'm worried inserting one is going to have very non-local effects that will be hard to reason about. It might be better if it would only act on the commands issued on the subgraph that it is attached to.

Reply:
> I'm worried inserting one is going to have very non-local effects that will be hard to reason about

This is a big part of why I want automatic inference.

Reply:
I'm not concerned.

We can easily detect and report:
- every pair of ambiguous nodes, both in the non-flattened sub-graphs and in the flattened system graph
- every edge between nodes that live under different sub-graphs (error if one or both sub-graphs are atomic)
- every edge between nodes that don't have build-time component access conflicts

So as a user, you would always be informed of ambiguities between `flush_commands` instances (and sub-graphs that contain them). I see no reason to complicate applying commands with all this graph stuff.

Reply:
> So as a user, you would always be informed of ambiguities between flush_commands instances (and sub-graphs that contain them).

Right. I agree with Joy here now.

Reply:
My concern is actually the opposite problem. That commands will be applied when the user doesn't expect it. For the most part this might not be a problem, since applying commands earlier shouldn't be wrong.  There will however be instances of multiple `flush_commands` in a exclusive systems pass. So some of those will be empty system runs. Again probably not a big deal, but feels a little weird.

One advantage of only acting on the sub graph would be that users could create their own systems for applying commands without having to worry about a bunch of side effects for commands that you're not expecting to handle. Probably not worth the extra complexity of implementation for now, but something to think about.

Reply:
What do you mean by "commands will be applied when the user doesn't expect them to be"?

Am I correct in interpreting this as "If I have two labels (subgraphs) that are ambiguous but both contain instances of `flush_commands`, what happens?"

Because that _was_ my point. The scheduler builder would warn you that "labels `A` and `B` are ambiguous but have conflicting access" or "label `X` and system `S` are ambiguous but have conflicting access." You wouldn't be blindsided. We could be extra vocal about it when `flush_commands` is involved.

If we add atomicity, labels should be "coherent" by default. I'd wager that allowing systems in different subgraphs to be interleaved is almost never idempotent.

Reply:
> Users could create their own systems for applying commands without having to worry about a bunch of side effects for commands that you're not expecting to handle.

We can't run subgraphs that contain sync points in parallel without side effects.

First, we can't (currently) know which archetypes a system can create/affect with its commands, so we have to assume that all commands produce non-local effects.

Second, in the absence of ambiguities, the single-threaded executor and multi-threaded executor must produce identical behavior. This is why labels should be "coherent" by default. Taking two sub-graphs and just mashing their systems together has a pretty high chance of violating that.

If you're sure that two subgraphs have no conflict except for their sync points, you should rework them into one subgraph with more appropriate sync points. You can always do this for your own logic. For plugins, we should just encourage plugin authors who want to export labels containing sync points to also export more granular labels that don't.

Reply:
> First, we can't (currently) know which archetypes a system can create/affect with its commands, so we have to assume that all commands produce non-local effects.

Agreed.

> Second, in the absence of ambiguities, the single-threaded executor and multi-threaded executor must produce identical behavior. 

This is a really interesting and powerful property. I very much like this.

> This is why labels should be "coherent" by default. Taking two sub-graphs and just mashing their systems together has a pretty high chance of violating that.

Interesting 🤔 "In the absence of ambiguities" is a very strong statement: I'm not convinced this actually follows as a result, but I am absolutely sympathetic to this line of thinking.

> For plugins, we should just encourage plugin authors who want to export labels containing sync points to also export more granular labels that don't.

I agree with this.

Reply:
> Interesting 🤔 "In the absence of ambiguities" is a very strong statement: I'm not convinced this actually follows as a result, but I am absolutely sympathetic to this line of thinking.

I'm pretty sure the current behavior follows this since both executors process commands queues in topological order. If there are no ambiguities, both executors will run writers (for the same component) in the same order and readers will appear between the same writers.

Reply:
> We could be extra verbose about it when flush_commands is involved (or make it an error).

Are all exclusive systems then ambiguous with each other if there's no defined order? I think that makes sense, but seems like it'd get very noisy in the ambiguity reporter.

> Users could create their own systems for applying commands without having to worry about a bunch of side effects for commands that you're not expecting to handle.

If a user tried to create a custom `flush_commands` system, they would have to create it very generically, because the flush commands system would potentially pull in commands from systems not in the subgraph. If it's guaranteed that the commands they have to deal with only come from the attached subgraph then it would be possible to tailor the behavior for just those commands.

Reply:
> Are all exclusive systems then ambiguous with each other if there's no defined order?

Yes, of course. Even in current Bevy, we report any pair of systems that *could* run in parallel but have conflicting access.

No false negatives (false positives if systems borrow same `ComponentId` but always different `ArchetypeComponentId`).

> I think that makes sense, but seems like it'd get very noisy in the ambiguity reporter.

If you're seeing ambiguities reported, you're probably doing something wrong. You don't need to have direct constraints between each pair of exclusive systems, but there must be a transitive order between them. Otherwise, depending on their dependencies and how quickly those finish, their order could change each iteration.

> If a user tried to create a custom `flush_commands` system...

If you say, "Trust me, the commands from subgraph `A` won't ever touch or create archetypes that subgraph `B` iterates," Bevy can't prove that. It's like `unsafe`. There are clear reasons (above) to flush all pending commands and I'm not seeing why would shouldn't always do that.

I think duplicating `Commands` as another system param that `flush_commands` doesn't touch is more reasonable.

Reply:
> If I read the atomic stuff correctly, "coherent" only releases locks when that wouldn't lead to observably different results, right?

Precisely.